ArrayList implementation:

Mwxyz2022 opened the pull request:
- Implemented full logic for all List interface methods
- Added internal dynamic array with initial capacity
- Implemented automatic resizing with ensureCapacity()
- Improved element shifting (right/left) during insert/remove
- Added index validation with custom exception
- Implemented remove by value and index
- Introduced get, set, size, isEmpty methods

Mwxyz2022 committed:
ArrayList implementation:
- Implemented full logic for all List interface methods
- Added internal dynamic array with initial capacity
- Implemented automatic resizing with ensureCapacity()
- Improved element shifting (right/left) during insert/remove
- Added index validation with custom exception
- Implemented remove by value and index
- Introduced get, set, size, isEmpty methods

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,152 @@
 
     @Override
     public void addAll(List<T> list) {
+        Object[] newArrayList = new Object[capacity + list.size()];
+
+        for (int i = 0; i < size; i++) {
+            newArrayList[i] = elements[i];
+        }
 
+        for (int i = size; i < size + list.size(); i++) {
+            newArrayList[i] = list.get(i - size);
+        }
+        elements = newArrayList;
+        capacity += list.size();
+        size += list.size();
```
Issue: The addAll method creates a new array with capacity + list.size(), but this does not follow the ensureCapacity logic used elsewhere. This could lead to inconsistent resizing and potential inefficiency. The checklist likely expects consistent resizing logic across all add methods. Consider using ensureCapacity to handle resizing, and then copy elements as needed.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,152 @@
+    }
+
+    private void shiftRightFromIndex(int index) {
+        for (int i = size; i >= index; i--) {
+            if (i == 0) {
+                elements[i] = null;
+                return;
+            }
+            elements[i] = elements[i - 1];
+        }
```
Issue: In shiftRightFromIndex, setting elements[0] = null when i == 0 is unnecessary and could cause data loss if shifting from index 0. The loop should simply shift elements right without nullifying elements[0]. This is a logic error.

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

Thank you for your work on the custom `ArrayList<T>` implementation! Your code covers most of the required functionality, but there are a couple of critical issues that need to be addressed before the solution can be approved:

1. **Logic Error in `shiftRightFromIndex`**:  
   - The method currently sets `elements[0] = null` when shifting elements, which is unnecessary and can cause data loss if shifting starts from index 0. The loop should only shift elements to the right without modifying `elements[0]`. Please fix this logic error to ensure no data is lost during shifting.

2. **Inconsistent Resizing in `addAll`**:  
   - The `addAll` method directly creates a new array with `capacity + list.size()`, bypassing the `ensureCapacity` logic used elsewhere in your class. This can lead to inconsistent resizing behavior and potential inefficiency. To maintain consistency and reliability, please use the `ensureCapacity` method before adding elements from the provided list, and then copy the elements as needed.

These issues are critical because they can lead to incorrect behavior and data loss, which means the core requirements of the task are not fully met yet. Once you address these points, your implementation should be ready for approval!

Keep up the good work—you're very close! If you need more information, you can check out [Java ArrayList source code](https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/util/ArrayList.java) or [Oracle's ArrayList documentation](https://docs.oracle.com/javase/8/docs/api/java/util/ArrayList.html) for reference. Let me know if you have any questions or need further clarification! 🚀

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

Mwxyz2022 committed:
- Changed ensureCapacity to no-arg method using internal size check
- Updated add, add at index, and addAll to use new ensureCapacity
- Simplified shiftRightFromIndex and shiftLeftFromIndex logic
- Removed redundant capacity parameter usage in resizing

mateacademy-ai-mentor reviewed (approved):
### Review Summary

Great job! 🎉 Your implementation meets all the core requirements for a custom ArrayList, and all previous issues have been addressed. The code is functional, follows the checklist, and no critical issues remain. Keep up the good work and continue practicing clean, robust code!

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

liliia-ponomarenko commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,143 @@
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private Object[] elements;
+    private int size = 0;
+    private int capacity = 10;
```
```suggestion
    private int size;
    private int capacity = DEFAULT_CAPACITY;
```

liliia-ponomarenko commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,143 @@
+        return size == 0;
+    }
+
+    private void shiftRightFromIndex(int index) {
+        for (int i = size; i >= index; i--) {
+            if (i > 0) {
+                elements[i] = elements[i - 1];
+            }
+        }
+    }
+
+    private void shiftLeftFromIndex(int index) {
+        for (int i = index; i < size - 1; i++) {
+            elements[i] = elements[i + 1];
+        }
+        elements[size - 1] = null;
+    }
```
Do you really need both methods? Can you replace it using `System.arraycopy`?

liliia-ponomarenko commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,143 @@
+
+    private void ensureCapacity(int minCapacity) {
+        if (minCapacity > capacity) {
+            while (capacity < minCapacity) {
+                capacity = capacity + (capacity >> 1);
+            }
+            Object[] temp = new Object[capacity];
+            for (int i = 0; i < size; i++) {
+                temp[i] = elements[i];
+            }
```
The same. System.arraycoppy increases the performance of your solution.

liliia-ponomarenko commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,143 @@
 
     @Override
     public void addAll(List<T> list) {
+        ensureCapacity(size + list.size());
+
+        for (int i = 0; i < list.size(); i++) {
+            elements[size + i] = list.get(i);
+        }
```
think how use method `add` here

liliia-ponomarenko commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,143 @@
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (element == null) {
+                if (elements[i] == null) {
+                    shiftLeftFromIndex(i);
+                    size--;
+                    return null;
+                }
+            } else if (element.equals(elements[i])) {
+                T removed = (T) elements[i];
+                shiftLeftFromIndex(i);
+                size--;
+                return removed;
+            }
```
move logic to get the element index to the private method and simplify it using `System.arraycopy`

liliia-ponomarenko reviewed (changes requested):
Good job! Let’s improve your solution;)

Mwxyz2022 committed:
Simplified addAll using add()
- moved index search to getElementIndex()
- used Objects.equals() for null-safe comparison
- used System.arraycopy() for shifting
- removed extra capacity field

Elena-Bruyako commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,132 @@
+
+    private int getElementIndex(T element) {
+        for (int i = 0; i < size; i++) {
+            if (Objects.equals(element, elements[i])) {
```
[Don't use Objects, Arrays, or any other util class.](https://github.com/mate-academy/jv-array-list/blob/master/checklist.md#dont-use-objects-arrays-or-any-other-util-class)

Elena-Bruyako commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,132 @@
+        return size == 0;
+    }
+
+    private void shiftRightFromIndex(int index) {
```
redundant separate method, you can use this logic directly in method

Elena-Bruyako commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,132 @@
     public T remove(T element) {
-        return null;
+        int index = getElementIndex(element);
+        T removeElement = (T) elements[index];
```
you can reuse method remove(index) here

Elena-Bruyako commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,132 @@
+        System.arraycopy(elements, index, elements, index + 1, size - index);
+    }
+
+    private void shiftLeftFromIndex(int index) {
```
you can use this logic directly in method

Mwxyz2022 committed:
Refactoring:
- removed Objects.equals() from getElementIndex()
- added manual null-safe comparison instead
- extracted right/left shift logic into shiftRightFromIndex() and shiftLeftFromIndex()
- updated add() and remove() to use new shift methods
- inlined remove(T element) to avoid duplicate logic

Mwxyz2022 committed:
Made removeElement final to fix Checkstyle warning
